src/packet/compressed_data.js

```diff
@@ -168,18 +166,45 @@ function node_zlib(func, create, options = {}) {
   };
 }
 
-function pako_zlib(constructor, options = {}) {
-  return function(data) {
-    const obj = new constructor(options);
-    return stream.transform(data, value => {
-      if (value.length) {
-        obj.push(value, Z_SYNC_FLUSH);
-        return obj.result;
-      }
-    }, () => {
-      if (constructor === Deflate) {
-        obj.push([], Z_FINISH);
-        return obj.result;
+function fflate_zlib(ZlibStreamedConstructor, options) {
+  return data => {
+    if (!util.isStream(data) || stream.isArrayStream(data)) {
+      return stream.fromAsync(() => stream.readToEnd(data).then(inputData => {
+        return new Promise((resolve, reject) => {
+          const zlibStream = new ZlibStreamedConstructor(options);
+          zlibStream.ondata = processedData => {
+            resolve(processedData);
+          };
+          try {
+            zlibStream.push(inputData, true); // only one chunk to push
+          } catch (err) {
+            reject(err);
+          }
+        });
+      }));
+    }
+
+    const inputReader = data.getReader();
+    const zlibStream = new ZlibStreamedConstructor(options);
+
+    return new ReadableStream({
+      async start(controller) {
+        zlibStream.ondata = async (value, isLast) => {
+          controller.enqueue(value);
+          if (isLast) {
+            controller.close();
+          }
+        };
+
+        while (true) {
```

Comment:
Since it seems that fflate calls `ondata` synchronously in all these cases (right?), you can actually handle backpressure without any extra effort by doing something like:
```suggestion
      }

      async pull() {
```
(and then the `pull` function will get called repeatedly until the internal queue is full).

Reply:
After discussing offline, we decided to leave this as future improvement, as the spec says (differing from MDN):

> Additionally, it will only be called repeatedly if it enqueues at least one chunk or fulfills a BYOB request; a no-op pull() implementation will not be continually called.